homeassistant/components/madvr/config_flow.py

```diff
@@ -0,0 +1,161 @@
+"""Config flow for the integration."""
+
+import asyncio
+import logging
+
+import aiohttp
+from madvr.madvr import HeartBeatError, Madvr
+import voluptuous as vol
+
+from homeassistant import config_entries
+from homeassistant.config_entries import ConfigEntry, ConfigFlowResult, OptionsFlow
+from homeassistant.const import CONF_HOST, CONF_MAC, CONF_NAME, CONF_PORT
+from homeassistant.core import callback
+
+from .const import DOMAIN
+from .coordinator import MadVRCoordinator
+from .utils import CannotConnect, cancel_tasks
+
+_LOGGER = logging.getLogger(__name__)
+
+
+class MadVRConfigFlow(config_entries.ConfigFlow, domain=DOMAIN):
+    """Handle a config flow for the integration."""
+
+    async def async_step_user(self, user_input=None) -> ConfigFlowResult:
+        """Handle the initial step."""
+        errors = {}
+
+        if user_input is not None:
+            host = user_input[CONF_HOST]
+            port = user_input.get(CONF_PORT, 44077)
+            mac = user_input[CONF_MAC]
+            keep_power_on = user_input["keep_power_on"]
+            try:
+                await self._test_connection(host, port, mac, keep_power_on)
+                return self.async_create_entry(
+                    title=user_input[CONF_NAME], data=user_input
+                )
+            except CannotConnect:
+                _LOGGER.error("CannotConnect error caught")
+                errors["base"] = "cannot_connect"
+
+        # Whether it's the first attempt or a retry, show the form
+        return self.async_show_form(
+            step_id="user",
+            data_schema=vol.Schema(
+                {
+                    vol.Required(
+                        CONF_NAME,
+                        default=user_input.get(CONF_NAME, "") if user_input else "",
+                    ): str,
+                    vol.Required(
+                        CONF_HOST,
+                        default=user_input.get(CONF_HOST, "") if user_input else "",
+                    ): str,
+                    vol.Required(
+                        CONF_MAC,
+                        default=user_input.get(CONF_MAC, "") if user_input else "",
+                    ): str,
+                    vol.Required(
+                        CONF_PORT,
+                        default=user_input.get(CONF_PORT, 44077)
+                        if user_input
+                        else 44077,
+                    ): int,
+                    vol.Required(
+                        "keep_power_on",
+                        default=user_input.get("keep_power_on", False)
+                        if user_input
+                        else False,
+                    ): bool,
+                }
```

Comment:
Please use `self.add_suggested_values` to add suggested values. Please also move the data schema out of the ConfigFlow. Check `pyload` for an example

Reply:
thank you for the example